app/models/message.rb

```diff
@@ -13,59 +14,45 @@ class Message < ActiveRecord::Base
   belongs_to :person
   belongs_to :to, class_name: 'Person', foreign_key: 'to_person_id'
   belongs_to :parent, class_name: 'Message', foreign_key: 'parent_id'
-  has_many :children, -> { where('to_person_id is null') }, class_name: 'Message', foreign_key: 'parent_id', dependent: :destroy
+  has_many :children, -> { where('to_person_id is null') },
+           class_name: 'Message', foreign_key: 'parent_id', dependent: :destroy
   has_many :attachments, dependent: :destroy
   has_many :log_items, -> { where(loggable_type: 'Message') }, foreign_key: 'loggable_id'
   belongs_to :site
+  has_and_belongs_to_many :members, class_name: 'Person', join_table: :members_messages, association_foreign_key: :member_id
 
   scope_by_site_id
-
-  scope :same_as, -> m { where('id != ?', m.id || 0).where(person_id: m.person_id, subject: m.subject, body: m.body, to_person_id: m.to_person_id, group_id: m.group_id).where('created_at >= ?', 1.day.ago) }
+  scope(:same_as, lambda do |m|
+    where('id != ?', m.id || 0)
+      .where(person_id:    m.person_id,
+             subject:      m.subject,
+             body:         m.body,
+             to_person_id: m.to_person_id,
+             group_id:     m.group_id)
+      .where('created_at >= ?', 1.day.ago)
+  end)
+  scope(:for_user, lambda do |user|
+    joins(:members).where('members_messages.member_id = ? OR messages.person_id = ?', user.id, user.id).uniq 
+  end)
+  scope :for_whole_group, -> { includes(:members).where(members_messages: { :message_id => nil }) }
 
   validates_presence_of :person_id
-  validates_presence_of :subject
   validates_length_of :subject, minimum: 2
-
-  validates_each :to_person_id, allow_nil: true do |record, attribute, value|
-    if attribute.to_s == 'to_person_id' and value and record.to and record.to.email.nil?
-      record.errors.add attribute, :invalid
-    end
-  end
-
-  validates_each :body do |record, attribute, value|
-    if attribute.to_s == 'body' and value.to_s.blank? and record.html_body.to_s.blank?
-      record.errors.add attribute, :blank
-    end
-  end
-
-  def top
-    top = self
-    while top.parent
-      top = top.parent
-    end
-    return top
+  validate do |message|
+    errors.add(:to, :invalid) if message.to && message.to.email.nil?
+    errors.add(:body, :blank) unless body || html_body
```

Comment:
You need to check `body.present? || html_body.present?` in case either are a blank string.